logpy/facts.py

```diff
@@ -29,24 +31,41 @@ def add_fact(self, *inputs):
             self.index[key].add(fact)
 
     def __call__(self, *args):
-        def f(s):
-            args2 = reify(args, s)
+        """ Returns an evaluated (callable) goal, which returns a list of
+        substitutions which match args against a fact in the knowledge base.
+
+        *args: the goal to evaluate. This consists of vars and values to
+               match facts against.
+
+        >>> from logpy.facts import Relation
+        >>> from logpy.variable import var
+        >>>
+        >>> x, y = var('x'), var('y')
+        >>> r = Relation()
+        >>> r.add_fact(1, 2, 3)
+        >>> r.add_fact(4, 5, 6)
+        >>> list(r(x, y, 3)({})) == [{y: 2, x: 1}]
+        True
+        >>> list(r(x, 5, y)({})) == [{y: 6, x: 4}]
+        True
+        >>> list(r(x, 42, y)({}))
+        []
+        """
+        def goal(substitution):
+            args2 = reify(args, substitution)
             subsets = [self.index[key] for key in enumerate(args)
                                        if  key in self.index]
             if subsets:     # we are able to reduce the pool early
                 facts = intersection(*sorted(subsets, key=len))
             else:
                 facts = self.facts
-            varinds = [i for i, arg in enumerate(args2) if isvar(arg)]
-            valinds = [i for i, arg in enumerate(args2) if not isvar(arg)]
-            vars = index(args2, varinds)
-            vals = index(args2, valinds)
-            assert not any(var in s for var in vars)
-
-            return (merge(dict(zip(vars, index(fact, varinds))), s)
-                              for fact in self.facts
-                              if vals == index(fact, valinds))
-        return f
+
+            for fact in facts:
+                unified = unify(fact, args2, substitution)
+                if unified != False:
```

Comment:
If this line is replaced with the simpler `if unified:`, then the behavior is incorrect in cases where no substitutions are required, eg:
```
>>> from logpy.facts import *
>>> from logpy.core import *
>>> r = Relation()
>>> r.add_fact(1, 2)
>>> run(0, x, r(1, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'x' is not defined
>>> x = var()
>>> run(0, x, r(1, 2))  # should be (x, )
()
```

This is covered by other portions of the test suite, but I can add an explicit test for it if you want.